言及機能を作成するの提出

YuSa0-6 opened the pull request:
- mention tableを作成し、組み合わせのユニーク制約を追加する
- reportとmentionとの関連付けを行う
- `reports/show.html.erb`に言及元の一覧を作成する
- レポート作成時にレポートのURIがあった場合は参照先としてMentionテーブルに保存する
- 更新時は全件削除し、再度関連する日報の言及先を保存する
- 更新、作成直後にMentionテーブルに登録するように設定。

YuSa0-6 committed:
Mention model作成

YuSa0-6 committed:
reportsとmentionsとのアソシエーションを追加

YuSa0-6 committed:
mentionsテーブルを定義。indexを使用してmentionテーブルの参照先ID、参照元IDの組み合わせが一意になるようにした。

YuSa0-6 committed:
ER図を確認したかったため

YuSa0-6 committed:
16行目17行目も含みます。
レポート作成時、URIのIDを取得し、Mentionテーブルに格納する。
レポート削除時、URIのIDを取得する
一旦関係のあるMentionテーブルを削除し、再度関係のあるMentionテーブルを作成する

YuSa0-6 committed:
参照された日報の一覧をリンクとして表示するように

YuSa0-6 committed:
reports/:idだけ許可するように

YuSa0-6 committed:
自分自身のレポートを言及しないようにした

YuSa0-6 committed:
必要ないのでコメントを削除する

YuSa0-6 commented on app/models/report.rb:
```diff
@@ -4,14 +4,59 @@ class Report < ApplicationRecord
   end
 
+  def other_report?(uri_id)
+    uri_id.to_i != id.to_i
```
そもそもこのメゾット自体必要かどうかは自分では自信を持って判断できませんでした。
ただ自分自身の日報を参照先として表示するのは違和感があったため、追記しました。

JunichiIto replied:
> 自分自身の日報を参照先として表示するのは違和感があった

そうですね、自分の日報は言及の対象外にするのがいいと思います。

ただ、uri_idはともかくidは最初から整数値なのでto_iするのはuri_idだけでいいですね。

あと、「メゾット」ではなく「メソッド」なのでご注意を〜。

YuSa0-6 commented on db/migrate/20240418061507_create_mentions.rb:
```diff
@@ -0,0 +1,13 @@
+class CreateMentions < ActiveRecord::Migration[7.0]
+  def change
+    create_table :mentions do |t|
+      t.integer :mentioning_report_id, null:false
+      t.integer :mentioned_report_id,  null:false
+      t.timestamps
+
+      t.index :mentioning_report_id
+      t.index :mentioned_report_id
+      t.index [:mentioning_report_id, :mentioned_report_id], unique: true
```
[テーブルの同じモデルの外部キーを複数もたせる](https://h-piiice16.hatenablog.com/entry/2018/08/13/114358)を参考にして試したが、上記の方が理解しやすかったのでこちらを利用した。参考としては会社の先輩に聞きました。

JunichiIto replied:
複合ユニークキー制約を持たせることと、外部キー制約を持たせることは別物です。
この書き方だと前者の制約は付きますが、後者は付きません。
外部キー制約も付けてみましょう。

YuSa0-6 commented on test/fixtures/mentions.yml:
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+# This model initially had no columns defined. If you add columns to the
+# model remove the "{}" from the fixture names and add the columns immediately
+# below each fixture, per the syntax in the comments below
+#
+one: {}
+# column: value
+#
+two: {}
+# column: value
```
こちらのファイルに関してはpushするかの判断ができなかったので、ご教示いただけると幸いです。

JunichiIto replied:
テストは現時点では考慮不要なので削除してしまって大丈夫です！

YuSa0-6 committed:
rubocop を通す

YuSa0-6 committed:
rubcoopを再度通す

JunichiIto commented on Gemfile:
```diff
@@ -56,6 +56,7 @@ group :development, :test do
 end
 
 group :development do
+  gem 'rails-erd'
```
提出物に関係ないgemは使ったら削除して元に戻しておく方がよいですー。

JunichiIto commented on app/models/mention.rb:
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class Mention < ApplicationRecord
+  # この言及元(mentioned_reports)は日報に対して一つだから
```
特にこのコメントがなくてもコードの可読性には影響がないかなーと思いました

JunichiIto commented on app/models/mention.rb:
```diff
@@ -0,0 +1,10 @@
+  # この言及先(mentioning_reports)は日報に対して１つだから
+  belongs_to :mentioning_report, class_name: 'Report', inverse_of: :mentioned_mentions
+
+  validates :mentioning_report_id, uniqueness: { scope: :mentioned_report_id, message: 'has already mentioned report' }
```
messageを指定していますが、このメッセージを表示する場面ってありました？
基本的にこのバリデーションは「念のため」であって、通常の操作ではエラーになる可能性は低い（ゆえに画面に表示されることもない）と思うので、デフォルトのメッセージのままで良いと思います。

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,59 @@ class Report < ApplicationRecord
   belongs_to :user
   has_many :comments, as: :commentable, dependent: :destroy
 
+  has_many :mentioning_mentions, class_name: 'Mention', foreign_key: :mentioning_report_id, dependent: :destroy # rubocop:disable Rails/InverseOf
```
`# rubocop:disable Rails/InverseOf`でrubocopの警告を無視した理由ってありますか？
特別な理由がない限りrubocopの警告は無視せずに解消してください〜。

YuSa0-6 replied:
[WIP]inverse_ofを使用すると期待通りの動きになりませんでした。

JunichiIto replied:
期待どおりに動きにならない、というのは具体的に何が起きたんでしょうか？
「コードを○○に変えたところ、○○のようになると思っていたが、実際は○○のような動きになった」というように、YuSaさんが実際に書いたコードや実際の実行結果などを交えて説明してほしいです〜。
（レビュアーが[エスパー](https://blog.jnito.com/entry/2020/04/17/072343)しなくて済むようにする）

YuSa0-6 replied:
# やりたいこと・わからないこと

### やりたいこと

- 例）
    - `Report.new(user_id: 1, title: "inverse_of testing", content: "http://localhost:3000/reports/1")` を実行してreports/60を作成した場合
        - Mentionテーブルには `mentioning_report_id: 1, mentioned_report_id: 60` が入っている状態にすること

### わからないこと

- Inverse_ofを使用すると上記のような動きではなく、 `mentioning_report_id: 60, mentioned_report_id: 60` のように同じレコードが入ってしまう状態です。まだ原因がわからないのでご教示いただければ幸いです。
# 実際にやったこと

- report.rbにて下記の通りにinverse_ofを記述しました。
    
    ```ruby
    # frozen_string_literal: true
    
    class Report < ApplicationRecord
      belongs_to :user
      has_many :comments, as: :commentable, dependent: :destroy
    
      has_many :mentioning_mentions, class_name: 'Mention', foreign_key: :mentioning_report_id,
                                     dependent: :destroy, inverse_of: :mentioning_report
      has_many :mentioned_reports, through: :mentioning_mentions, source: :mentioned_report
    
      has_many :mentioned_mentions, class_name: 'Mention', foreign_key: :mentioned_report_id,
                                    dependent: :destroy, inverse_of: :mentioning_report
      has_many :mentioning_reports, through: :mentioned_mentions, source: :mentioned_report
    
      validates :title, presence: true
      validates :content, presence: true
    
      after_save :update_mentions
    
      def editable?(target_user)
        user == target_user
      end
    
      def other_report?(uri_id)
        uri_id.to_i != id
      end
    
      def created_on
        created_at.to_date
      end
    
      def update_mentions
        uri_ids = extract_uri_ids
        mentioned_reports.destroy_all
        save_mentions(uri_ids)
      end
    
      def save_mentions(uri_ids)
        uri_ids.each do |uri_id|
          next unless other_report?(uri_id)
    
          # Mention.new(mentioned_report_id: uri_id, mentioning_report_id: id).save
          mentioned_mentions.new(mentioning_report_id: uri_id).save!
        end
      end
    
      def extract_uri_ids
        uris = URI.extract(content)
        uris.map do |uri|
          uri.match(%r{http://localhost:3000/reports/(\d+)$}).captures.first
        end
      end
    end
    
    ```
    

# 実際に試した結果

rails コンソールにてReport作成→Mention.allで確認してみました。

```ruby
report_record = Report.new(user_id: 1, title: "inverse_of testing", content: "http://localhost:3000/reports/1").save!
  TRANSACTION (0.1ms)  begin transaction
  User Load (0.2ms)  SELECT "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  Report Create (0.8ms)  INSERT INTO "reports" ("user_id", "title", "content", "created_at", "updated_at") VALUES (?, ?, ?, ?, ?)  [["user_id", 1], ["title", "inverse_of testing"], ["content", "http://localhost:3000/reports/1"], ["created_at", "2024-05-17 03:12:51.694393"], ["updated_at", "2024-05-17 03:12:51.694393"]]
  Report Load (0.0ms)  SELECT "reports".* FROM "reports" INNER JOIN "mentions" ON "reports"."id" = "mentions"."mentioned_report_id" WHERE "mentions"."mentioning_report_id" = ?  [["mentioning_report_id", 62]]
  Report Load (0.0ms)  SELECT "reports".* FROM "reports" WHERE "reports"."id" = ? LIMIT ?  [["id", 62], ["LIMIT", 1]]
  Mention Exists? (0.1ms)  SELECT 1 AS one FROM "mentions" WHERE "mentions"."mentioning_report_id" = ? AND "mentions"."mentioned_report_id" = ? LIMIT ?  [["mentioning_report_id", 62], ["mentioned_report_id", 62], ["LIMIT", 1]]
  Mention Create (0.2ms)  INSERT INTO "mentions" ("mentioning_report_id", "mentioned_report_id", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["mentioning_report_id", 62], ["mentioned_report_id", 62], ["created_at", "2024-05-17 03:12:51.698338"], ["updated_at", "2024-05-17 03:12:51.698338"]]
  TRANSACTION (1.5ms)  commit transaction
=> true
```

```ruby
irb(main):008:0> Mention.all
  Mention Load (0.2ms)  SELECT "mentions".* FROM "mentions"
=> [#<Mention:0x00000001088aa3c8 id: 1, mentioning_report_id: 62, mentioned_report_id: 62, created_at: Fri, 17 May 2024 12:12:51.698338000 JST +09:00, updated_at: Fri, 17 May 2024 12:12:51.698338000 JST +09:00>]
```

結果として何故か同じID62が入ってしまいます。

UIとしては自分を参照してしまっている状態です。

# 原因だとおもわれること

関連付けにおいてInverse_ofで関連付けの優先順位が変わっている気がする。（？）
- 道中考えたこと
  - テーブルの定義は正しいのか？
  - 関連付けは正しいのか？
  - uri_idsは正しく取得できているのか？


# 参考
- [[[ActiveRecord] 双方向関連付けとinverse_of](https://zenn.dev/igaiga/books/rails-practice-note/viewer/ar_inverse_of)](https://zenn.dev/igaiga/books/rails-practice-note/viewer/ar_inverse_of)
- [[Active Record の関連付け - Railsガイド](https://railsguides.jp/association_basics.html#%E5%8F%8C%E6%96%B9%E5%90%91%E9%96%A2%E9%80%A3%E4%BB%98%E3%81%91)](https://railsguides.jp/association_basics.html#双方向関連付け)

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,59 @@ class Report < ApplicationRecord
   validates :title, presence: true
   validates :content, presence: true
 
+  after_save :extract_uri_ids_save_mentions
```
英語的に自然なのは `extract_uri_ids_and_save_mentions` でしょうか。
ただ、ちょっと具体的過ぎるのでもうちょっと抽象度を上げてシンプルな名前にした方がいいかもしれません。
https://bootcamp.fjord.jp/articles/78

YuSa0-6 replied:
ここはメソッド名を決める時に結構迷いました。
可能であれば抽象度を上げたいのですが、どれを取捨選択するべきかを考えていました。
今回のメインはmentionを更新・作成することなので`save_mentions`や`update_mentions`でも良い気がしてきました。

JunichiIto replied:
updateは「更新のみ？（新規作成はしない？）」と勘違いされるかもしれないのでsaveの方がいいかもしれませんね

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,59 @@ class Report < ApplicationRecord
   validates :content, presence: true
 
+  after_save :extract_uri_ids_save_mentions
+  after_update :extract_uri_ids_update_mentions
```
saveとupdateで分けていますが、saveだとcreateもupdateも対象になるので、似たような処理が2回実行されることになりますー。

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,59 @@ class Report < ApplicationRecord
+    uri_ids.each do |uri_id|
+      next unless other_report?(uri_id)
+
+      Mention.new(mentioned_report_id: uri_id, mentioning_report_id: id).save
```
関連をうまく使うとベターです。以下の記事は取得の時を対象にしていますが、保存も同じ考え方ができます。
https://bootcamp.fjord.jp/articles/76

あと、`save`メソッドは万一保存に失敗しても無視されるので、`save!`のように失敗したら例外が出るようにした方がベターです。

JunichiIto replied:
これは対応していない？

YuSa0-6 replied:
申し訳ありません、変更を反映できてなかったです。

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,59 @@ class Report < ApplicationRecord
+  end
+
+  def delete_mentions
+    Mention.where(mentioning_report_id: id).find_each(&:destroy)
```
ここも関連を使いましょう。
あと、`find_each(&:destroy)`のかわりに`destroy_all`が使えます。

また、これぐらいのコードならわざわざメソッド化しなくてもいいかなと思いました。
https://bootcamp.fjord.jp/pages/when-you-should-define-method

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,59 @@ class Report < ApplicationRecord
+  def extract_uri_ids
+    uris = URI.extract(content)
+    uris.map do |uri|
+      uri.match(%r{reports/(\d+)$}).captures.first
```
この正規表現だと全く無関係なドメインのURLもマッチしてしまうので、http://localhost:3000 だけが対象になるようにしてください。

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,59 @@ class Report < ApplicationRecord
+    uris = URI.extract(content)
+    uris.map do |uri|
+      uri.match(%r{reports/(\d+)$}).captures.first
+    rescue StandardError
```
この例外処理はどういうときに機能しますか？
「ここでこういう場合に例外が発生する」と事前にわかってる場合は例外処理ではなく、条件分岐等を使うのがベターです。
チェリー本を持っていたら、例外処理のベストプラクティスという項を読んでみてください〜。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -4,8 +4,15 @@
 
+<strong>mentioned_reports</strong>
+<div class="index-items">
+  <% @report.mentioned_reports.each do |report| %>
```
orderを指定した方が良いです。
https://qiita.com/jnchito/items/90c28c3f3e856add0a82

JunichiIto reviewed (commented):
コメントしました〜。

YuSa0-6 committed:
ユニーク制約だけでなく、外部キー制約も追加する。

YuSa0-6 committed:
使用しないメッセージを削除する。

YuSa0-6 committed:
外部キーの追加

YuSa0-6 committed:
外部キー制約を追加するためにt.referencesに変更する。

YuSa0-6 committed:
[review] https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1582614709
提出時に関係ないGemなので削除しておく。

YuSa0-6 committed:
[review] https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1582615230
可読性に影響がないのでコメントを削除する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1582617346
rubocopの指示通りにinverse_ofを追記して、関連先のモデルを設定する。

YuSa0-6 committed:
メソッドの抽象度を下げ、同じ処理を二回走らないようにするために`extract_uri_ids_update_mentions`を削除して`update_mentions`を利用する。

YuSa0-6 committed:
[review] https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1582619677
`id` は元から整数値なので`to_i`を削除する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1582625210
メソッド化する必要性が低いので、解体して１行に記述した。destroy_allを削除時に使用した。

YuSa0-6 committed:
orderを追記して、並び順を固定する

YuSa0-6 committed:
特定のドメインが対象になるように正規表現を修正

YuSa0-6 committed:
テストは現時点で使用しないので削除する

YuSa0-6 committed:
関連付けでレコード登録できるようにする。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1604139040
更新と、新規作成で使用するメソッドのな名前変更。updateよりsaveの方がメソッド名としては良いと判断したので変更する。

YuSa0-6 committed:
typo修正する

YuSa0-6 committed:
型を変更する

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,16 +4,18 @@ class Report < ApplicationRecord
-  has_many :mentioning_mentions, class_name: 'Mention', foreign_key: :mentioning_report_id, dependent: :destroy, inverse_of: :mentioned_report
+  has_many :mentioning_mentions, class_name: 'Mention', foreign_key: :mentioning_report_id,
+                                 dependent: :destroy
   has_many :mentioned_reports, through: :mentioning_mentions, source: :mentioned_report
```
mentioning_mentions が「自分から送った言及」の意味であれば、この関連は「自分が言及している日報」という意味で mentioning_reports と名付けるのが自然な気がします。

mentioned_reportsだと受け身形なので「自分が他から言及されている日報」という意味に見えるんですが、どうでしょうか？

このあたりの命名や実装を見直した上で、もう一度inverse_ofを試してみてほしいです〜。
うまくいかなかったらまた詳細を教えてください。

YuSa0-6 replied:
@JunichiIto 
伊藤さんの説明通りであっています。
- 言及している日報-> `mentioning_mentions'
- 言及された日報-> 'mentioned_mentions'
## 困っていること・質問
どう手をつけていいか全く見当がついていないのですが、
- `has_many :mentioning_reports, class_name: 'Mention', foreign_key: :mentioning_report_id,
                                 dependent: :destroy, inverse_of: mentioning_reports`
とした場合は`has_many :mentioned_reports, through: mentioning_reports ....` と言うふうに記述する認識であっていますか？
- この場合、mentioning_reportsの名前が被っても問題ないでしょうか？
理解が足りず申し訳ないですが、ご教示いただけると幸いです。

YuSa0-6 replied:
# 解決しました（？）
## ためしたこと
- 中間テーブルである、mentions.rbのInverse_ofを削除して、関連付けを見直しました。

JunichiIto commented on app/models/report.rb:
```diff
@@ -27,17 +29,19 @@ def created_on
+  def create_mentions(uri_ids)
+    transaction do
+      uri_ids.each do |uri_id|
+        next unless other_report?(uri_id)
```
```suggestion
        next if uri_id.to_i == id
```

これでも「自分に対する言及だったら処理をスキップ」と読めるのでメソッド化はしなくても大丈夫かもしれません。

JunichiIto commented on app/models/report.rb:
```diff
@@ -27,17 +29,19 @@ def created_on
   end
 
-  def update_mentions
+  def save_mentions
```
どのメソッドがpublicでどのメソッドがprivateなのか、メソッドの可視性を検討してみてください。

JunichiIto reviewed (commented):
コメントしました！

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1605658789
メソッド化しなくても処理内容が明白なのでother_report?メソッドを削除する。

YuSa0-6 commented on app/models/report.rb:
```diff
@@ -4,14 +4,51 @@ class Report < ApplicationRecord
   def created_on
     created_at.to_date
   end
+
```
private入れる。
コントローラーで使うものはpublick

YuSa0-6 reviewed (commented):
Inverse_ofの件多分解決しました。
中間テーブルにもInverse_ofを記述していたのが原因だと考えてます。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1605658553
- mention.rbのinverse_ofを削除する
- report.rbの関連名を見直しする
- save_mentoinsのメソッドを関連付けに合わせて変更する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1605658823
- 判断基準は以下の通りにしました。
	- controllerやviewで呼び出していないか
-public以外はすべてprivate

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,13 +4,12 @@ class Report < ApplicationRecord
-  has_many :mentioned_mentions, class_name: 'Mention', foreign_key: :mentioned_report_id,
-                                dependent: :destroy
-  has_many :mentioning_reports, through: :mentioned_mentions, source: :mentioning_report
+  has_many :mentions_from, class_name: 'Mention', foreign_key: :mentioned_report_id, dependent: :destroy,
```
Railsのルール的に、最後に複数形の"s"で終わる命名の方がトラブルが避けやすいと思います。
たとえば、`active_mentions` / `passive_mentions` みたいな名前はどうでしょうか？

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,48 @@ class Report < ApplicationRecord
+  private
+
+  def save_mentions
+    uri_ids = extract_uri_ids
```
変数の寿命を短くしましょう
https://bootcamp.fjord.jp/pages/311

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,48 @@ class Report < ApplicationRecord
+  end
+
+  def create_mentions(uri_ids)
+    transaction do
```
after_saveのコールバックはそれ自体でtransactionがかかっている（saveメソッド全体が1つのtransactionになっている）ので、ここは無くしちゃっても大丈夫です

JunichiIto commented on app/models/report.rb:
```diff
@@ -4,14 +4,48 @@ class Report < ApplicationRecord
+      uri_ids.each do |uri_id|
+        next if uri_id.to_i == id
+
+        mentions_from.new(mentioning_report_id: uri_id).save!
```
```suggestion
        mentions_from.create!(mentioning_report_id: uri_id)
```

でもいけそう

JunichiIto reviewed (commented):
コメントしました！

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1610823546
関連付けの名前を変更した。
Railsルール的に複数形のsで終わる方がトラブルを避けやすいらしい。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1610824256
after_save( saveメソッド自体がtransaction)になっているみたいなので不要のため削除する。

YuSa0-6 committed:
[review]https://github.com/YuSa0-6/fjord-books_app-2023/pull/6#discussion_r1610823949
変数の寿命を短くする。

JunichiIto reviewed (approved):
確認しました！OKです〜🙆‍♂️
